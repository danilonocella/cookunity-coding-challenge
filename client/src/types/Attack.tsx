export type Attack = {
  "_id": number,
  "name": string,
  "damage": number
}
