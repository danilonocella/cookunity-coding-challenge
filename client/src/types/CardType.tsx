export type CardType =
  | "Grass"
  | "Fire"
  | "Water"
  | "Electric"
  | "Psychic"
  | "Fighting"
  | "Darkness"
  | "Metal"
  | "Dragon"
  | "Fairy"
  | "Colorless";
