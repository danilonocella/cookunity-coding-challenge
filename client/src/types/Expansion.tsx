export type Expansion =
  | "Base set"
  | "Jungle"
  | "Fossil"
  | "Base set 2"
  | "Team rocket"
  | "Gym heroes"
  | "Gym challenge";
