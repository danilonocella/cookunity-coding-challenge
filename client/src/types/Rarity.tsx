export type Rarity = "Common" | "Uncommon" | "Rare";
