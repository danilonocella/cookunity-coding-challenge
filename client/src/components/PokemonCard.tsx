import React, { useEffect, useState } from "react";
import { Card as CardComponent, Typography, CardMedia, Box, Grid, Stack } from "@mui/material";
import { styled } from "@mui/system";
import imagesServiceInstance from "../apis/imagesServiceConfig";
import { getRarityIcon } from "../utils/getRarityIcon";
import { getTypeIcon } from "../utils/getTypeIcon";
import { Colors, Card } from "../types";

interface PokemonCardProps {
  card: Card;
  onClick?: () => void;
}

const StyledCard = styled(CardComponent)(({ color, onClick }) => ({
  width: 325,
  border: "2px solid #000",
  borderRadius: 10,
  backgroundColor: color,
  boxShadow: "0 4px 8px rgba(0, 0, 0, 0.1)",
  overflow: "hidden",
  fontFamily: "Arial, sans-serif",
  padding: 10,
  position: "relative",
  cursor: onClick && "pointer",
}));

export const CardHeader = styled(Box)({
  display: "flex",
  justifyContent: "space-between",
  alignItems: "center",
  padding: 4,
  backgroundColor: "#fff",
  borderRadius: 5,
  border: "2px solid #000",
});

const HitPoints = styled(Typography)({
  color: "#ff0000",
  fontWeight: "bold",
  fontSize: 18,
});

const CardImage = styled(Box)({
  borderRadius: 5,
  overflow: "hidden",
  border: "2px solid #000",
  margin: "10px 0",
  padding: 44,
  backgroundColor: "#fff",
});

const CardAttack = styled(Grid)({
  padding: 16,
  marginTop: 40,
  marginBottom: 40,
});

const Footer = styled(Grid)({
  marginTop: 10,
  backgroundColor: "#fff",
  borderRadius: 5,
  border: "2px solid #000",
  padding: 4,
});

const AttackText = styled(Typography)({
  fontWeight: 700,
  fontSize: 18,
});

const PokemonCard: React.FC<PokemonCardProps> = ({ card, onClick }) => {
  const [imageUrl, setImageUrl] = useState<string>("");

  useEffect(() => {
    const fetchCardImage = async () => {
      const lowercaseName = card.name.toLowerCase();
      try {
        const response = await imagesServiceInstance.get(`/${lowercaseName}`);
        setImageUrl(response.data.sprites.front_default);
      } catch (error) {
        console.error(error);
      }
    };

    fetchCardImage();
  }, [card.name]);

  return (
    <StyledCard
      color={Colors[card.type as keyof typeof Colors]}
      onClick={onClick}
      data-testid="pokemon-card"
    >
      <CardHeader data-testid="card-header">
        <Typography variant="h6" component="div">
          {card.name}
        </Typography>
        <Stack direction="row" alignItems="center" gap={1}>
          <HitPoints variant="h6">{card.hit_points} HP</HitPoints>
          {getTypeIcon(card.type)}
        </Stack>
      </CardHeader>
      <CardImage data-testid="card-image">
        <CardMedia
          component="img"
          height="200"
          image={imageUrl}
          alt={card.name}
        />
      </CardImage>
      <Grid>
        <CardAttack item container xs={12} justifyContent="space-between" data-testid="card-attack">
          <Grid item>
            <AttackText color="textSecondary">{card.attack.name}</AttackText>
          </Grid>
          <Grid item>
            <AttackText color="textSecondary">{card.attack.damage}</AttackText>
          </Grid>
        </CardAttack>
        <Footer item container justifyContent="space-between" data-testid="card-footer">
          <Grid
            item
            xs={4}
            container
            alignItems="center"
            justifyContent="flex-start"
          >
            <Stack direction="row" alignItems="center">
              <Typography variant="caption">Weakness:</Typography>
              {getTypeIcon(card.weakness.weakness_type)}
            </Stack>
          </Grid>
          <Grid
            item
            xs={4}
            container
            alignItems="center"
            justifyContent="center"
          >
            <Stack direction="row" alignItems="center">
              <Typography variant="caption">Resistance:</Typography>
              {getTypeIcon(card.resistance_type)}
            </Stack>
          </Grid>
          <Grid
            item
            xs={4}
            container
            alignItems="center"
            justifyContent="flex-end"
          >
            <Stack direction="row" alignItems="center">
              <Typography variant="caption">Rarity:</Typography>
              {getRarityIcon(card.rarity)}
            </Stack>
          </Grid>
        </Footer>
      </Grid>
    </StyledCard>
  );
};

export default PokemonCard;
