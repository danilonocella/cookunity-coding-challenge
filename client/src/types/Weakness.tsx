import { CardType } from "./CardType"

export type Weakness = {
  "_id": number,
  "main_type": CardType,
  "weakness_type": CardType
}
