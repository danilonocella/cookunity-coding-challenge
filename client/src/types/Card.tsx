import { Attack } from "./Attack";
import { CardType } from "./enums/CardType";
import { Expansion } from "./enums/Expansion";
import { Rarity } from "./Rarity";
import { Weakness } from "./Weakness";

export type Card = {
  id: number;
  name: string;
  type: CardType;
  rarity: Rarity;
  expansion: Expansion;
  hit_points: number;
  resistance_type: CardType;
  resistance_points: number;
  attack: Attack;
  weakness: Weakness;
};
