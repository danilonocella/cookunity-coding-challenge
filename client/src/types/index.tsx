import { Rarity } from "./Rarity";
import { Weakness } from "./Weakness"
import { Attack } from "./Attack"
import {Card } from './Card'

export type { Rarity, Weakness, Attack, Card };
export * from "./enums/CardType";
export * from "./enums/Expansion";
export * from "./enums/Colors";
