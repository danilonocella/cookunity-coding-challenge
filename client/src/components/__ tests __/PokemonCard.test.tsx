import { render, screen, waitFor } from "@testing-library/react";
import PokemonCard from "../PokemonCard";
import { Card } from "../../types";
import imagesServiceInstance from "../../apis/imagesServiceConfig";

jest.mock("../../apis/imagesServiceConfig");
const mockedImagesServiceInstance = imagesServiceInstance as jest.Mocked<
  typeof imagesServiceInstance
>;

const mockCard: Card = {
  id: 4,
  name: "Pikachu",
  type: "Electric",
  rarity: "Common",
  expansion: "Base set",
  hit_points: 40,
  resistance_type: "Colorless",
  resistance_points: 20,
  attack: {
    _id: 4,
    name: "Gnaw",
    damage: 20,
  },
  weakness: {
    _id: 4,
    main_type: "Electric",
    weakness_type: "Fighting",
  },
};

const mockImageUrl =
  "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png";

describe("PokemonCard", () => {
  beforeEach(() => {
    mockedImagesServiceInstance.get.mockResolvedValue({
      data: {
        sprites: {
          front_default: mockImageUrl,
        },
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("matches the snapshot", async () => {
    const { asFragment } = render(<PokemonCard card={mockCard} />);

    await waitFor(() =>
      expect(mockedImagesServiceInstance.get).toHaveBeenCalledTimes(1)
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the PokemonCard component", async () => {
    render(<PokemonCard card={mockCard} />);

    await waitFor(() =>
      expect(mockedImagesServiceInstance.get).toHaveBeenCalledTimes(1)
    );

    expect(screen.getByTestId("pokemon-card")).toBeInTheDocument();
    expect(screen.getByTestId("card-header")).toBeInTheDocument();
    expect(screen.getByTestId("card-image")).toBeInTheDocument();
    expect(screen.getByTestId("card-attack")).toBeInTheDocument();
    expect(screen.getByTestId("card-footer")).toBeInTheDocument();
    expect(screen.getByText("Pikachu")).toBeInTheDocument();
    expect(screen.getByText("40 HP")).toBeInTheDocument();
    expect(screen.getByText("Gnaw")).toBeInTheDocument();
    expect(screen.getByText("20")).toBeInTheDocument();
    expect(screen.getByAltText("Pikachu")).toHaveAttribute("src", mockImageUrl);
  });
});
