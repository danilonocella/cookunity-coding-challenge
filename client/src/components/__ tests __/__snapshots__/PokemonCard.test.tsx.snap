// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1mb0zpg-MuiPaper-root-MuiCard-root"
    color="#FCD022"
    data-testid="pokemon-card"
  >
    <div
      class="MuiBox-root css-3cwqp8"
      data-testid="card-header"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
      >
        Pikachu
      </div>
      <div
        class="MuiStack-root css-1rlitzi-MuiStack-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-nfz5wh-MuiTypography-root"
        >
          40 HP
        </h6>
        <svg
          aria-hidden="true"
          aria-label="Electric"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-gp4nhn-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="type-Electric-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14.69 2.21 4.33 11.49c-.64.58-.28 1.65.58 1.73L13 14l-4.85 6.76c-.22.31-.19.74.08 1.01.3.3.77.31 1.08.02l10.36-9.28c.64-.58.28-1.65-.58-1.73L11 10l4.85-6.76c.22-.31.19-.74-.08-1.01-.3-.3-.77-.31-1.08-.02"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiBox-root css-4e9ocd"
      data-testid="card-image"
    >
      <img
        alt="Pikachu"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
        height="200"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png"
      />
    </div>
    <div
      class="MuiGrid-root css-vj1n65-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-g7nv63-MuiGrid-root"
        data-testid="card-attack"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-125vsvh-MuiTypography-root"
          >
            Gnaw
          </p>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-125vsvh-MuiTypography-root"
          >
            20
          </p>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item css-2u5hxj-MuiGrid-root"
        data-testid="card-footer"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-4 css-102ghzi-MuiGrid-root"
        >
          <div
            class="MuiStack-root css-1d9cypr-MuiStack-root"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              Weakness:
            </span>
            <svg
              aria-hidden="true"
              aria-label="Fighting"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1xqx3iu-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="type-Fighting-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 20c0 .55.45 1 1 1h8c.55 0 1-.45 1-1v-3H7zM18 7c-.55 0-1 .45-1 1V5c0-1.1-.9-2-2-2H7c-1.1 0-2 .9-2 2v5.8c0 .13.01.26.04.39l.8 4c.09.47.5.8.98.8h10.36c.45 0 .89-.36.98-.8l.8-4c.03-.13.04-.26.04-.39V8c0-.55-.45-1-1-1m-3 3H7V7h8z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-4 css-1qzfdba-MuiGrid-root"
        >
          <div
            class="MuiStack-root css-1d9cypr-MuiStack-root"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              Resistance:
            </span>
            <svg
              aria-hidden="true"
              aria-label="Colorless"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-dbcmfq-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="type-Colorless-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-4 css-fiumd8-MuiGrid-root"
        >
          <div
            class="MuiStack-root css-1d9cypr-MuiStack-root"
          >
            <span
              class="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
            >
              Rarity:
            </span>
            <svg
              aria-hidden="true"
              aria-label="Common"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1n3zrz1-MuiSvgIcon-root"
              data-mui-internal-clone-element="true"
              data-testid="rarity-common-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
